File: src/engine/state.ts
/**
 * Durable run state. The graph must be able to answer, at any moment:
 *   What has already happened?  (events.jsonl, node records)
 *   Why did the system choose this route?  (decisions[] with the state that produced them)
 *   Where can execution safely resume?  (state.json checkpoint after every node)
 *
 * Layout: runs/<run_id>/
 *   run.json         run record (spec snapshot, input, status, totals, approvals, decisions)
 *   state.json       node records (checkpoint)
 *   events.jsonl     append-only event log
 *   artifacts/       full prompts/raw outputs per attempt (state stores references, not blobs)
 *   inbox/           tasks for the orchestrator bridge (+ results)
 *   approvals/       human gate decisions
 *   nested/<node>/   loop rounds and subgraph runs (each a full run dir)
 */
import fs from "node:fs";
import path from "node:path";
import crypto from "node:crypto";
import type { Budget, GraphSpec, NodeKind } from "../spec/schema.js";
import { sumUsage, type Usage } from "../models.js";

export type RunStatus = "created" | "running" | "paused" | "completed" | "failed" | "cancelled";
export type NodeStatus = "pending" | "running" | "waiting_approval" | "waiting_task" | "completed" | "failed" | "skipped";

export interface AttemptRecord {
  attempt: number;
  item?: number;
  started_at: string;
  ended_at?: string;
  duration_ms?: number;
  model?: string;
  bridge?: string;
  status: "ok" | "error" | "invalid_output" | "timeout" | "cancelled";
  error?: string;
  validation_errors?: string[];
  cost_usd?: number;
  usage?: Usage;
  artifact?: string;
  source?: string;
}

export interface ItemRecord {
  index: number;
  status: "pending" | "running" | "completed" | "failed";
  output?: unknown;
  error?: string;
  attempts: number;
  duration_ms?: number;
  cost_usd?: number;
}

export interface KilledRecord {
  index: number;
  item: unknown;
  reasons: string[];
  confidence: number;
}

export interface NodeRecord {
  id: string;
  kind: NodeKind;
  status: NodeStatus;
  started_at?: string;
  ended_at?: string;
  duration_ms?: number;
  output?: unknown;
  /** Map / verify fan-out per-item records. */
  items?: ItemRecord[];
  /** Successful item outputs in order. */
  outputs?: unknown[];
  attempts: AttemptRecord[];
  cost_usd: number;
  usage: Usage;
  agent_calls: number;
  retries: number;
  repairs: number;
  error?: string;
  skip_reason?: string;
  route?: string;
  route_reason?: string;
  verify?: { total: number; survivors: unknown[]; killed: KilledRecord[]; kill_rate: number; repair_round: number };
  gate?: { decision?: "approved" | "rejected" | "auto"; by?: string; comment?: string; at?: string; requested_at?: string };
  loop?: { rounds: number; stop_reason: string; converged: boolean; collected: number; seen: number; cost_usd: number };
  /** Set when a verify repair cycle re-runs this node. */
  repair?: { round: number; feedback: unknown };
  side_effect_done?: boolean;
  /** Which attempt produced `output`. */
  artifact?: string;
  waiting_task?: string;
}

export interface ApprovalRecord {
  gate: string;
  decision: "approved" | "rejected";
  by: string;
  comment?: string;
  at: string;
}

export interface Decision {
  at: string;
  node: string;
  kind: "route" | "gate" | "loop" | "verify" | "failure" | "budget" | "skip" | "repair";
  detail: string;
  state?: Record<string, unknown>;
}

export interface RunTotals {
  cost_usd: number;
  usage: Usage;
  agent_calls: number;
  retries: number;
  wall_ms: number;
}

export interface RunRecord {
  id: string;
  graph: string;
  spec_file: string;
  spec: GraphSpec;
  input: unknown;
  status: RunStatus;
  created_at: string;
  updated_at: string;
  started_at?: string;
  ended_at?: string;
  bridge: string;
  budget: Budget;
  frozen: string[];
  error?: string;
  output?: unknown;
  totals: RunTotals;
  approvals: Record<string, ApprovalRecord>;
  decisions: Decision[];
  parent?: { run_id: string; node_id: string; round?: number };
  labels?: Record<string, string>;
  /** Extra scope for nested runs (e.g. { loop: { round, seen, collected, prev } }). */
  context?: Record<string, unknown>;
  /** Visible degradation: nodes that failed but were tolerated, etc. */
  warnings?: string[];
  /** Set when this run was created by `gren fork` from another run. */
  forked_from?: string;
}

export interface RunState {
  run: RunRecord;
  nodes: Record<string, NodeRecord>;
}

export interface GrenEvent {
  ts: string;
  seq: number;
  run_id: string;
  type: string;
  node?: string;
  item?: number;
  data?: Record<string, unknown>;
}

export interface InboxTask {
  task_id: string;
  run_id: string;
  node_id: string;
  item_index?: number;
  attempt: number;
  role: "agent" | "verify";
  model: string;
  effort?: string;
  system?: string;
  prompt: string;
  output_schema: Record<string, unknown>;
  tools?: string[];
  created_at: string;
  timeout_at: string;
  status: "pending" | "claimed" | "completed" | "cancelled";
  claimed_by?: string;
  claimed_at?: string;
  instructions: string;
}

export interface InboxResult {
  task_id: string;
  output: unknown;
  cost_usd?: number;
  usage?: Usage;
  model?: string;
  source?: "orchestrator" | "human" | "subagent" | "api" | "mock";
  worker?: string;
  completed_at: string;
  error?: string;
}

export interface RunSummary {
  id: string;
  graph: string;
  status: RunStatus;
  created_at: string;
  updated_at: string;
  cost_usd: number;
  bridge: string;
  nodes: { total: number; completed: number; failed: number; skipped: number; waiting: number };
  parent?: RunRecord["parent"];
}

export function nowIso(): string {
  return new Date().toISOString();
}

export function newRunId(prefix = "run"): string {
  const d = new Date();
  const stamp = d.toISOString().replace(/[-:]/g, "").replace(/\..+/, "").replace("T", "-");
  return `${prefix}-${stamp}-${crypto.randomBytes(3).toString("hex")}`;
}

function atomicWrite(file: string, data: string) {
  const tmp = `${file}.${process.pid}.${crypto.randomBytes(2).toString("hex")}.tmp`;
  fs.writeFileSync(tmp, data, "utf8");
  fs.renameSync(tmp, file);
}

export class RunStore {
  private seq = new Map<string, number>();

  constructor(public readonly root: string) {
    fs.mkdirSync(root, { recursive: true });
  }

  runDir(runId: string): string {
    // nested runs are addressed as "<parent>/nested/<node>/<child>" - keep path semantics
    return path.join(this.root, runId);
  }

  exists(runId: string): boolean {
    return fs.existsSync(path.join(this.runDir(runId), "run.json"));
  }

  create(opts: {
    id?: string;
    spec: GraphSpec;
    specFile: string;
    input: unknown;
    bridge: string;
    budget: Budget;
    frozen: string[];
    parent?: RunRecord["parent"];
    labels?: Record<string, string>;
  }): RunState {
    const id = opts.id ?? newRunId(opts.spec.name.replace(/[^a-zA-Z0-9]+/g, "-").slice(0, 24));
    const dir = this.runDir(id);
    fs.mkdirSync(path.join(dir, "artifacts"), { recursive: true });
    fs.mkdirSync(path.join(dir, "inbox"), { recursive: true });
    fs.mkdirSync(path.join(dir, "approvals"), { recursive: true });
    const ts = nowIso();
    const run: RunRecord = {
      id,
      graph: opts.spec.name,
      spec_file: opts.specFile,
      spec: opts.spec,
      input: opts.input,
      status: "created",
      created_at: ts,
      updated_at: ts,
      bridge: opts.bridge,
      budget: opts.budget,
      frozen: opts.frozen,
      totals: { cost_usd: 0, usage: {}, agent_calls: 0, retries: 0, wall_ms: 0 },
      approvals: {},
      decisions: [],
      parent: opts.parent,
      labels: opts.labels,
    };
    const nodes: Record<string, NodeRecord> = {};
    for (const n of opts.spec.nodes) {
      nodes[n.id] = { id: n.id, kind: n.kind, status: "pending", attempts: [], cost_usd: 0, usage: {}, agent_calls: 0, retries: 0, repairs: 0 };
    }
    const state: RunState = { run, nodes };
    this.save(state);
    return state;
  }

  load(runId: string): RunState {
    const dir = this.runDir(runId);
    const runFile = path.join(dir, "run.json");
    if (!fs.existsSync(runFile)) throw new Error(`run not found: ${runId}`);
    const run = JSON.parse(fs.readFileSync(runFile, "utf8")) as RunRecord;
    const stateFile = path.join(dir, "state.json");
    const nodes = fs.existsSync(stateFile) ? (JSON.parse(fs.readFileSync(stateFile, "utf8")) as { nodes: Record<string, NodeRecord> }).nodes : {};
    return { run, nodes };
  }

  save(state: RunState) {
    state.run.updated_at = nowIso();
    const dir = this.runDir(state.run.id);
    fs.mkdirSync(dir, { recursive: true });
    atomicWrite(path.join(dir, "run.json"), JSON.stringify(state.run, null, 2));
    atomicWrite(path.join(dir, "state.json"), JSON.stringify({ nodes: state.nodes }, null, 2));
  }

  appendEvent(runId: string, evt: Omit<GrenEvent, "ts" | "seq" | "run_id">): GrenEvent {
    const seq = (this.seq.get(runId) ?? this.lastSeq(runId)) + 1;
    this.seq.set(runId, seq);
    const full: GrenEvent = { ts: nowIso(), seq, run_id: runId, ...evt };
    fs.appendFileSync(path.join(this.runDir(runId), "events.jsonl"), JSON.stringify(full) + "\n", "utf8");
    return full;
  }

  private lastSeq(runId: string): number {
    const f = path.join(this.runDir(runId), "events.jsonl");
    if (!fs.existsSync(f)) return 0;
    const lines = fs.readFileSync(f, "utf8").trim().split("\n").filter(Boolean);
    if (!lines.length) return 0;
    try {
      return (JSON.parse(lines[lines.length - 1]!) as GrenEvent).seq ?? lines.length;
    } catch {
      return lines.length;
    }
  }

  readEvents(runId: string, afterSeq = 0): GrenEvent[] {
    const f = path.join(this.runDir(runId), "events.jsonl");
    if (!fs.existsSync(f)) return [];
    const out: GrenEvent[] = [];
    for (const line of fs.readFileSync(f, "utf8").split("\n")) {
      if (!line.trim()) continue;
      try {
        const e = JSON.parse(line) as GrenEvent;
        if (e.seq > afterSeq) out.push(e);
      } catch {
        /* ignore torn line */
      }
    }
    return out;
  }

  writeArtifact(runId: string, name: string, data: unknown): string {
    const rel = path.join("artifacts", `${name}.json`);
    const abs = path.join(this.runDir(runId), rel);
    fs.mkdirSync(path.dirname(abs), { recursive: true });
    atomicWrite(abs, JSON.stringify(data, null, 2));
    return rel.replace(/\\/g, "/");
  }

  readArtifact(runId: string, rel: string): unknown {
    const abs = path.join(this.runDir(runId), rel);
    if (!fs.existsSync(abs)) return undefined;
    return JSON.parse(fs.readFileSync(abs, "utf8"));
  }

  // ---- approvals (human gates) ----
  writeApproval(runId: string, approval: ApprovalRecord) {
    atomicWrite(path.join(this.runDir(runId), "approvals", `${approval.gate}.json`), JSON.stringify(approval, null, 2));
  }

  readApproval(runId: string, gate: string): ApprovalRecord | undefined {
    const f = path.join(this.runDir(runId), "approvals", `${gate}.json`);
    if (!fs.existsSync(f)) return undefined;
    try {
      return JSON.parse(fs.readFileSync(f, "utf8")) as ApprovalRecord;
    } catch {
      return undefined;
    }
  }

  // ---- inbox (orchestrator bridge) ----
  writeTask(task: InboxTask) {
    atomicWrite(path.join(this.runDir(task.run_id), "inbox", `${task.task_id}.json`), JSON.stringify(task, null, 2));
  }

  readTask(runId: string, taskId: string): InboxTask | undefined {
    const f = path.join(this.runDir(runId), "inbox", `${taskId}.json`);
    if (!fs.existsSync(f)) return undefined;
    return JSON.parse(fs.readFileSync(f, "utf8")) as InboxTask;
  }

  listTasks(runId: string): InboxTask[] {
    const dir = path.join(this.runDir(runId), "inbox");
    if (!fs.existsSync(dir)) return [];
    return fs
      .readdirSync(dir)
      .filter((f) => /^[^.]+\.json$/.test(f))
      .map((f) => {
        try {
          return JSON.parse(fs.readFileSync(path.join(dir, f), "utf8")) as InboxTask;
        } catch {
          return undefined;
        }
      })
      .filter((t): t is InboxTask => Boolean(t && typeof t.task_id === "string" && typeof t.created_at === "string"))
      .map((t) => (t.status !== "completed" && fs.existsSync(path.join(dir, `${t.task_id}.result.json`)) ? { ...t, status: "completed" as const } : t))
      .sort((a, b) => a.created_at.localeCompare(b.created_at));
  }

  /** Tasks for a run and all of its nested runs (loop rounds, subgraphs). */
  listTasksDeep(runId: string, onlyPending = true): InboxTask[] {
    const out: InboxTask[] = [];
    const walk = (id: string, depth: number) => {
      if (depth > 6) return;
      for (const t of this.listTasks(id)) if (!onlyPending || t.status === "pending" || t.status === "claimed") out.push(t);
      const nested = path.join(this.runDir(id), "nested");
      if (!fs.existsSync(nested)) return;
      for (const node of fs.readdirSync(nested, { withFileTypes: true })) {
        if (!node.isDirectory()) continue;
        const nodeDir = path.join(nested, node.name);
        for (const child of fs.readdirSync(nodeDir, { withFileTypes: true })) {
          if (child.isDirectory() && fs.existsSync(path.join(nodeDir, child.name, "run.json"))) walk(`${id}/nested/${node.name}/${child.name}`, depth + 1);
        }
      }
    };
    walk(runId, 0);
    return out;
  }

  /** All pending tasks across every run in the store. */
  listAllPendingTasks(): InboxTask[] {
    const out: InboxTask[] = [];
    for (const r of this.list()) {
      if (r.parent) continue; // nested runs are reached through their parent
      if (r.status !== "running" && r.status !== "paused" && r.status !== "created") continue;
      out.push(...this.listTasksDeep(r.id));
    }
    return out;
  }

  writeTaskResult(runId: string, result: InboxResult) {
    atomicWrite(path.join(this.runDir(runId), "inbox", `${result.task_id}.result.json`), JSON.stringify(result, null, 2));
  }

  readTaskResult(runId: string, taskId: string): InboxResult | undefined {
    const f = path.join(this.runDir(runId), "inbox", `${taskId}.result.json`);
    if (!fs.existsSync(f)) return undefined;
    try {
      return JSON.parse(fs.readFileSync(f, "utf8")) as InboxResult;
    } catch {
      return undefined;
    }
  }

  // ---- listing ----
  list(): RunSummary[] {
    const out: RunSummary[] = [];
    const walk = (dir: string, prefix: string, depth: number) => {
      if (!fs.existsSync(dir) || depth > 4) return;
      for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
        if (!entry.isDirectory()) continue;
        const id = prefix ? `${prefix}/${entry.name}` : entry.name;
        const full = path.join(dir, entry.name);
        if (fs.existsSync(path.join(full, "run.json"))) {
          try {
            out.push(summarize(this.load(id)));
          } catch {
            /* skip corrupt */
          }
          walk(path.join(full, "nested"), `${id}/nested`, depth + 1);
        } else if (entry.name === "nested" || prefix.endsWith("nested")) {
          walk(full, id, depth + 1);
        }
      }
    };
    walk(this.root, "", 0);
    return out.sort((a, b) => b.created_at.localeCompare(a.created_at));
  }

  delete(runId: string) {
    fs.rmSync(this.runDir(runId), { recursive: true, force: true });
  }

  /** Copy a run (state, artifacts, approvals, nested runs - not the inbox) under a new id. */
  fork(srcId: string, newId?: string): RunState {
    const src = this.load(srcId);
    const id = newId ?? newRunId(`${src.run.graph.replace(/[^a-zA-Z0-9]+/g, "-").slice(0, 20)}-fork`);
    const from = this.runDir(srcId);
    const to = this.runDir(id);
    if (fs.existsSync(to)) throw new Error(`run ${id} already exists`);
    fs.mkdirSync(to, { recursive: true });
    for (const sub of ["artifacts", "approvals", "nested"]) {
      if (fs.existsSync(path.join(from, sub))) fs.cpSync(path.join(from, sub), path.join(to, sub), { recursive: true });
    }
    fs.mkdirSync(path.join(to, "inbox"), { recursive: true });
    const state: RunState = { run: { ...src.run, id, status: "created", ended_at: undefined, error: undefined, output: undefined, warnings: undefined, forked_from: srcId, created_at: nowIso() }, nodes: src.nodes };
    this.save(state);
    this.appendEvent(id, { type: "run.forked", data: { from: srcId } });
    return state;
  }
}

export function summarize(state: RunState): RunSummary {
  const ns = Object.values(state.nodes);
  return {
    id: state.run.id,
    graph: state.run.graph,
    status: state.run.status,
    created_at: state.run.created_at,
    updated_at: state.run.updated_at,
    cost_usd: state.run.totals.cost_usd,
    bridge: state.run.bridge,
    nodes: {
      total: ns.length,
      completed: ns.filter((n) => n.status === "completed").length,
      failed: ns.filter((n) => n.status === "failed").length,
      skipped: ns.filter((n) => n.status === "skipped").length,
      waiting: ns.filter((n) => n.status === "waiting_approval" || n.status === "waiting_task").length,
    },
    parent: state.run.parent,
  };
}

export function addUsage(rec: { usage: Usage; cost_usd: number; agent_calls: number }, usage: Usage | undefined, cost: number | undefined, calls = 1) {
  rec.usage = sumUsage(rec.usage, usage);
  rec.cost_usd += cost ?? 0;
  rec.agent_calls += calls;
}

File: gren/gren/engine/state.py
"""Durable run state. The graph must be able to answer, at any moment:
  What has already happened?  (events.jsonl, node records)
  Why did the system choose this route?  (decisions[] with the state that produced them)
  Where can execution safely resume?  (state.json checkpoint after every node + the Strands session)

Layout (unchanged from gren 1 so the dashboard reads both): runs/<run_id>/
  run.json  state.json  events.jsonl  artifacts/  inbox/  approvals/  nested/<node>/<child>/  session/ (Strands)
"""
from __future__ import annotations

import json
import os
import random
import shutil
import time
from dataclasses import dataclass
from typing import Any

from ..models.pricing import sum_usage
from ..spec.schema import GraphSpec

TERMINAL = ("completed", "failed", "skipped")


def now_iso() -> str:
    t = time.time()
    return time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime(t)) + f".{int((t % 1) * 1000):03d}Z"


def new_run_id(prefix: str = "run") -> str:
    stamp = time.strftime("%Y%m%d-%H%M%S", time.gmtime())
    return f"{prefix}-{stamp}-{random.randbytes(3).hex()}"


def atomic_write(path: str, data: Any) -> None:
    """Write-then-rename with EPERM retries (Windows: the dashboard may be reading the file)."""
    os.makedirs(os.path.dirname(path), exist_ok=True)
    tmp = f"{path}.{os.getpid()}.{random.randbytes(2).hex()}.tmp"
    with open(tmp, "w", encoding="utf8") as f:
        json.dump(data, f, indent=2, default=str)
    for attempt in range(6):
        try:
            os.replace(tmp, path)
            return
        except PermissionError:
            time.sleep(0.015 * (attempt + 1))
    with open(path, "w", encoding="utf8") as f:
        json.dump(data, f, indent=2, default=str)
    try:
        os.remove(tmp)
    except OSError:
        pass


def read_json(path: str) -> Any:
    last: Exception | None = None
    for attempt in range(5):
        try:
            with open(path, encoding="utf8") as f:
                return json.load(f)
        except FileNotFoundError:
            raise
        except (json.JSONDecodeError, OSError) as e:
            last = e
            time.sleep(0.01 * (attempt + 1))
    raise last  # type: ignore[misc]


def new_node_record(node_id: str, kind: str) -> dict[str, Any]:
    return {"id": node_id, "kind": kind, "status": "pending", "attempts": [], "cost_usd": 0.0, "usage": {}, "agent_calls": 0, "retries": 0, "repairs": 0}


@dataclass
class RunState:
    run: dict[str, Any]
    nodes: dict[str, dict[str, Any]]


def add_usage(rec: dict[str, Any], usage: dict[str, Any] | None, cost: float | None, calls: int = 1) -> None:
    rec["usage"] = sum_usage(rec.get("usage"), usage)
    rec["cost_usd"] = float(rec.get("cost_usd", 0.0)) + float(cost or 0.0)
    rec["agent_calls"] = int(rec.get("agent_calls", 0)) + calls


class RunStore:
    def __init__(self, root: str):
        self.root = os.path.abspath(root)
        os.makedirs(self.root, exist_ok=True)
        self._seq: dict[str, int] = {}

    def run_dir(self, run_id: str) -> str:
        return os.path.join(self.root, *run_id.split("/"))

    def exists(self, run_id: str) -> bool:
        return os.path.exists(os.path.join(self.run_dir(run_id), "run.json"))

    def create(self, spec: GraphSpec, spec_file: str, input_: Any, bridge: str, budget: dict[str, Any], frozen: list[str], run_id: str | None = None,
               parent: dict[str, Any] | None = None, labels: dict[str, str] | None = None, context: dict[str, Any] | None = None) -> RunState:
        rid = run_id or new_run_id("".join(c if c.isalnum() else "-" for c in spec.name)[:24])
        d = self.run_dir(rid)
        for sub in ("artifacts", "inbox", "approvals"):
            os.makedirs(os.path.join(d, sub), exist_ok=True)
        ts = now_iso()
        run = {
            "id": rid, "graph": spec.name, "spec_file": spec_file, "spec": spec.model_dump(exclude_none=True, by_alias=True), "input": input_,
            "status": "created", "created_at": ts, "updated_at": ts, "bridge": bridge, "budget": budget, "frozen": frozen,
            "totals": {"cost_usd": 0.0, "usage": {}, "agent_calls": 0, "retries": 0, "wall_ms": 0}, "approvals": {}, "decisions": [],
            "parent": parent, "labels": labels, "context": context, "engine": "strands",
        }
        nodes = {n.id: new_node_record(n.id, n.kind) for n in spec.nodes}
        state = RunState(run=run, nodes=nodes)
        self.save(state)
        return state

    def load(self, run_id: str) -> RunState:
        d = self.run_dir(run_id)
        run_file = os.path.join(d, "run.json")
        if not os.path.exists(run_file):
            raise FileNotFoundError(f"run not found: {run_id}")
        run = read_json(run_file)
        state_file = os.path.join(d, "state.json")
        nodes = read_json(state_file).get("nodes", {}) if os.path.exists(state_file) else {}
        return RunState(run=run, nodes=nodes)

    def spec_of(self, state: RunState) -> GraphSpec:
        return GraphSpec.model_validate(state.run["spec"])

    def save(self, state: RunState) -> None:
        state.run["updated_at"] = now_iso()
        d = self.run_dir(state.run["id"])
        atomic_write(os.path.join(d, "run.json"), state.run)
        atomic_write(os.path.join(d, "state.json"), {"nodes": state.nodes})

    # ---- events ----
    def _last_seq(self, run_id: str) -> int:
        f = os.path.join(self.run_dir(run_id), "events.jsonl")
        if not os.path.exists(f):
            return 0
        last = 0
        with open(f, encoding="utf8") as fh:
            for line in fh:
                if line.strip():
                    try:
                        last = int(json.loads(line).get("seq", last))
                    except json.JSONDecodeError:
                        pass
        return last

    def append_event(self, run_id: str, type_: str, node: str | None = None, item: int | None = None, data: dict[str, Any] | None = None) -> dict[str, Any]:
        seq = self._seq.get(run_id, self._last_seq(run_id)) + 1
        self._seq[run_id] = seq
        evt = {"ts": now_iso(), "seq": seq, "run_id": run_id, "type": type_, "node": node, "item": item, "data": data or {}}
        d = self.run_dir(run_id)
        os.makedirs(d, exist_ok=True)
        with open(os.path.join(d, "events.jsonl"), "a", encoding="utf8") as f:
            f.write(json.dumps(evt, default=str) + "\n")
        return evt

    def read_events(self, run_id: str, after_seq: int = 0) -> list[dict[str, Any]]:
        f = os.path.join(self.run_dir(run_id), "events.jsonl")
        if not os.path.exists(f):
            return []
        out = []
        with open(f, encoding="utf8") as fh:
            for line in fh:
                if not line.strip():
                    continue
                try:
                    e = json.loads(line)
                except json.JSONDecodeError:
                    continue
                if e.get("seq", 0) > after_seq:
                    out.append(e)
        return out

    def nested_run_ids(self, run_id: str, depth: int = 0) -> list[str]:
        out: list[str] = []
        if depth > 6:
            return out
        nested = os.path.join(self.run_dir(run_id), "nested")
        if not os.path.isdir(nested):
            return out
        for node in sorted(os.listdir(nested)):
            node_dir = os.path.join(nested, node)
            if not os.path.isdir(node_dir):
                continue
            for child in sorted(os.listdir(node_dir)):
                if os.path.exists(os.path.join(node_dir, child, "run.json")):
                    cid = f"{run_id}/nested/{node}/{child}"
                    out.append(cid)
                    out.extend(self.nested_run_ids(cid, depth + 1))
        return out

    def read_events_deep(self, run_id: str) -> list[dict[str, Any]]:
        evs = self.read_events(run_id)
        for cid in self.nested_run_ids(run_id):
            evs.extend(self.read_events(cid))
        return sorted(evs, key=lambda e: e.get("ts", ""))

    # ---- artifacts ----
    def write_artifact(self, run_id: str, name: str, data: Any) -> str:
        rel = f"artifacts/{name}.json"
        atomic_write(os.path.join(self.run_dir(run_id), "artifacts", f"{name}.json"), data)
        return rel

    def read_artifact(self, run_id: str, rel: str) -> Any:
        p = os.path.join(self.run_dir(run_id), *rel.split("/"))
        return read_json(p) if os.path.exists(p) else None

    # ---- approvals ----
    def write_approval(self, run_id: str, approval: dict[str, Any]) -> None:
        atomic_write(os.path.join(self.run_dir(run_id), "approvals", f"{approval['gate']}.json"), approval)

    def read_approval(self, run_id: str, gate: str) -> dict[str, Any] | None:
        p = os.path.join(self.run_dir(run_id), "approvals", f"{gate}.json")
        try:
            return read_json(p) if os.path.exists(p) else None
        except Exception:
            return None

    def clear_approval(self, run_id: str, gate: str) -> None:
        p = os.path.join(self.run_dir(run_id), "approvals", f"{gate}.json")
        if os.path.exists(p):
            os.remove(p)

    # ---- inbox ----
    def _inbox(self, run_id: str) -> str:
        return os.path.join(self.run_dir(run_id), "inbox")

    def write_task(self, task: dict[str, Any]) -> None:
        atomic_write(os.path.join(self._inbox(task["run_id"]), f"{task['task_id']}.json"), task)

    def read_task(self, run_id: str, task_id: str) -> dict[str, Any] | None:
        p = os.path.join(self._inbox(run_id), f"{task_id}.json")
        return read_json(p) if os.path.exists(p) else None

    def list_tasks(self, run_id: str) -> list[dict[str, Any]]:
        d = self._inbox(run_id)
        if not os.path.isdir(d):
            return []
        out = []
        for f in os.listdir(d):
            if not f.endswith(".json") or f.count(".") != 1:
                continue
            try:
                t = read_json(os.path.join(d, f))
            except Exception:
                continue
            if not isinstance(t, dict) or "task_id" not in t or "created_at" not in t:
                continue
            if t.get("status") != "completed" and os.path.exists(os.path.join(d, f"{t['task_id']}.result.json")):
                t = {**t, "status": "completed"}
            out.append(t)
        return sorted(out, key=lambda t: t["created_at"])

    def list_tasks_deep(self, run_id: str, only_pending: bool = True) -> list[dict[str, Any]]:
        out = []
        for rid in [run_id, *self.nested_run_ids(run_id)]:
            for t in self.list_tasks(rid):
                if not only_pending or t.get("status") in ("pending", "claimed"):
                    out.append(t)
        return out

    def list_all_pending_tasks(self) -> list[dict[str, Any]]:
        out = []
        for r in self.list():
            if r.get("parent") or r["status"] not in ("running", "paused", "created"):
                continue
            out.extend(self.list_tasks_deep(r["id"]))
        return out

    def write_task_result(self, run_id: str, result: dict[str, Any]) -> None:
        atomic_write(os.path.join(self._inbox(run_id), f"{result['task_id']}.result.json"), result)

    def read_task_result(self, run_id: str, task_id: str) -> dict[str, Any] | None:
        p = os.path.join(self._inbox(run_id), f"{task_id}.result.json")
        try:
            return read_json(p) if os.path.exists(p) else None
        except Exception:
            return None

    # ---- listing / lifecycle ----
    def list(self) -> list[dict[str, Any]]:
        out: list[dict[str, Any]] = []

        def walk(d: str, prefix: str, depth: int) -> None:
            if not os.path.isdir(d) or depth > 4:
                return
            for entry in sorted(os.listdir(d)):
                full = os.path.join(d, entry)
                if not os.path.isdir(full):
                    continue
                rid = f"{prefix}/{entry}" if prefix else entry
                if os.path.exists(os.path.join(full, "run.json")):
                    try:
                        out.append(summarize(self.load(rid)))
                    except Exception:
                        pass
                    walk(os.path.join(full, "nested"), f"{rid}/nested", depth + 1)
                elif entry == "nested" or prefix.endswith("nested"):
                    walk(full, rid, depth + 1)

        walk(self.root, "", 0)
        return sorted(out, key=lambda r: r["created_at"], reverse=True)

    def delete(self, run_id: str) -> None:
        shutil.rmtree(self.run_dir(run_id), ignore_errors=True)

    def fork(self, src_id: str, new_id: str | None = None) -> RunState:
        src = self.load(src_id)
        rid = new_id or new_run_id("".join(c if c.isalnum() else "-" for c in src.run["graph"])[:20] + "-fork")
        frm, to = self.run_dir(src_id), self.run_dir(rid)
        if os.path.exists(to):
            raise FileExistsError(f"run {rid} already exists")
        os.makedirs(to, exist_ok=True)
        for sub in ("artifacts", "approvals", "nested"):
            if os.path.isdir(os.path.join(frm, sub)):
                shutil.copytree(os.path.join(frm, sub), os.path.join(to, sub))
        os.makedirs(os.path.join(to, "inbox"), exist_ok=True)
        run = {**src.run, "id": rid, "status": "created", "ended_at": None, "error": None, "output": None, "warnings": None, "forked_from": src_id, "created_at": now_iso()}
        state = RunState(run=run, nodes=src.nodes)
        self.save(state)
        self.append_event(rid, "run.forked", data={"from": src_id})
        return state


def summarize(state: RunState) -> dict[str, Any]:
    ns = list(state.nodes.values())
    r = state.run
    return {
        "id": r["id"], "graph": r["graph"], "status": r["status"], "created_at": r["created_at"], "updated_at": r["updated_at"],
        "cost_usd": r["totals"]["cost_usd"], "bridge": r["bridge"], "parent": r.get("parent"),
        "labels": r.get("labels") or None, "wall_ms": r["totals"].get("wall_ms"), "ended_at": r.get("ended_at"), "forked_from": r.get("forked_from"),
        "error": (r.get("error") or None) if r["status"] == "failed" else None,
        "nodes": {
            "total": len(ns),
            "completed": sum(1 for n in ns if n["status"] == "completed"),
            "failed": sum(1 for n in ns if n["status"] == "failed"),
            "skipped": sum(1 for n in ns if n["status"] == "skipped"),
            "waiting": sum(1 for n in ns if n["status"] in ("waiting_approval", "waiting_task")),
        },
    }
